### PR TITLE
[SPARK-46958][SQL] Add missing timezone to coerce default values 

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Add missing timezone for `Cast`s in ResolveDefaultColumn
- For the canUpcast branch, if the timezone is missing, INTERNAL_ERROR will be reported after the DDL succeeds cuz' the Cast is `unresolved.`
```java
[info] - SPARK-46958: Fix bug when canUpcast and result non-foldable expression *** FAILED *** (36 milliseconds)
[info]   org.apache.spark.SparkException: [INTERNAL_ERROR] Found the unresolved Cast: cast(2018-11-17 00:00:00 as string) SQLSTATE: XX000
[info] == SQL (line 1, position 1) ==
[info] INSERT INTO t (key) VALUES(1)
[info] ^^^^^^^^^^^^^^^^^^^
```
- For the `else` branch, where we cast a wider type to a narrow one for convenience when the value is fittable, it fails the DDL directly if the timezone is missing.

```java
[info] - SPARK-46958: timestamp should have timezone for resolvable when default values fit *** FAILED *** (5 milliseconds)
[info]   org.apache.spark.sql.AnalysisException: [INVALID_DEFAULT_VALUE.DATA_TYPE] Failed to execute CREATE TABLE command because the destination column or variable `c` has a DEFAULT value '2018-11-17', which requires "TIMESTAMP" type, but the statement provided a value of incompatible "STRING" type. SQLSTATE: 42623
```
### Why are the changes needed?

bugfix

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new test
### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/ResolveDefaultColumnsSuite.scala [hunk 1]
```diff
@@ -256,4 +256,11 @@ class ResolveDefaultColumnsSuite extends QueryTest with SharedSparkSession {
       checkAnswer(sql("select * from t"), Row(1, "apache", "spark "))
     }
   }
+
+  test("SPARK-46958: Fix bug when canUpcast and result non-foldable expression") {
+    sql(s"CREATE TABLE t(key int, c STRING DEFAULT date '2018-11-17') " +
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ResolveDefaultColumnsSuite.scala:261]: does it work if the default expression is `CAST(date '2018-11-17' AS STRING)`? I think we should find the root cause and understand why the casted expression was considered as non-foldable.

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/ResolveDefaultColumnsSuite.scala:261]: The case for the date type seems to be restored on the master branch, but ts values still have an issue caused by missing timeZone, thus, unresolved.

File: sql/core/src/test/scala/org/apache/spark/sql/ResolveDefaultColumnsSuite.scala [hunk 2]
```diff
@@ -182,14 +182,14 @@ class ResolveDefaultColumnsSuite extends QueryTest with SharedSparkSession {
             "actualType" -> "\"STRING\""))
         checkError(
           exception = intercept[AnalysisException] {
-            sql("create table demos.test_ts_other (a timestamp default '2022-01-02') using parquet")
```
Comment (yaooqinn) [inline sql/core/src/test/scala/org/apache/spark/sql/ResolveDefaultColumnsSuite.scala:185]: The original test seems incorrect as the expression `a timestamp default '2022-01-02'` is actually valid, which meets the requirement of fitting within the narrower typ

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala [hunk 3]
```diff
@@ -311,36 +313,35 @@ object ResolveDefaultColumns extends QueryErrorsBase
...
+        equivalent
+      case canUpCast if Cast.canUpCast(canUpCast.dataType, supplanted) =>
+        Cast(analyzed, supplanted, Some(conf.sessionLocalTimeZone))
+      case l: Literal
+        if !Seq(supplanted, l.dataType).exists(_ match {
           case _: BooleanType | _: ArrayType | _: StructType | _: MapType => true
           case _ => false
-        })) {
-        try {
-          val casted = Cast(analyzed, supplanted, evalMode = EvalMode.TRY).eval()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:329]: shall we make the change smaller and only add the missing timezone here?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala:329]: The syntax using match-case here is more elegant than if-else. Although it increases the LOC, it's limited to a single method, so it's safe from the perspective of `git diff`. However, if you insist, I can change it.

**Review by cloud-fan — APPROVED** [review]

**Comment by yaooqinn** [comment]
Thanks @cloud-fan, merged to master

**Pull request closed without merging**